src/test/java/io/confluent/connect/jdbc/sink/JdbcDbWriterTest.java

```diff
@@ -71,6 +80,59 @@ private JdbcDbWriter newWriter(Map<String, String> props) {
     return new JdbcDbWriter(config, dialect, dbStructure);
   }
 
+  private JdbcDbWriter newConnectionMockedWriter(Map<String, String> props, Connection mockConnection) {
+    final JdbcSinkConfig config = new JdbcSinkConfig(props);
+    dialect = mock(DatabaseDialect.class);
+    final DbStructure dbStructure = mock(DbStructure.class);
+    return new JdbcDbWriter(config, dialect, dbStructure) {
+      protected CachedConnectionProvider connectionProvider(int maxConnAttempts, long retryBackoff) {
+        CachedConnectionProvider mockConnectionProvider = mock(CachedConnectionProvider.class);
+        when(mockConnectionProvider.getConnection()).thenReturn(mockConnection);
+        return mockConnectionProvider;
+      }
+    };
+  }
+
+  @Test (expected = SQLException.class)
+  public void verifyConnectionRollback() throws SQLException {
+    Connection mockConnection = mock(Connection.class);
+
+    doThrow(new SQLException()).when(mockConnection).commit();
+
+    Map<String, String> props = new HashMap<>();
+    props.put("connection.url", sqliteHelper.sqliteUri());
+    props.put("auto.create", "true");
+    props.put("auto.evolve", "true");
+    props.put("insert.mode", "upsert");
+    props.put("pk.mode", "record_key");
+    props.put("pk.fields", "id"); // assigned name for the primitive key
+
+    JdbcDbWriter writer = newConnectionMockedWriter(props, mockConnection);
+
+    PreparedStatement mockStatement = mock(PreparedStatement.class);
+    when(dialect.parseTableIdentifier(any())).thenReturn(mock(TableId.class));
+    when(dialect.createPreparedStatement(any(), any())).thenReturn(mockStatement);
+    when(dialect.statementBinder(any(), any(), any(), any(), any(), any()))
+        .thenReturn(mock(PreparedStatementBinder.class));
+    when(mockStatement.executeBatch()).thenReturn(new int[3]);
+
+    Schema keySchema = Schema.INT64_SCHEMA;
+
+    Schema valueSchema1 = SchemaBuilder.struct()
+        .field("author", Schema.STRING_SCHEMA)
+        .field("title", Schema.STRING_SCHEMA)
+        .build();
+
+    Struct valueStruct1 = new Struct(valueSchema1)
+        .put("author", "Tom Robbins")
+        .put("title", "Villa Incognito");
+
+    writer.write(Collections.singleton(new SinkRecord("books", 0, keySchema, 1L, valueSchema1,
+        valueStruct1, 0)));
+
+    verify(mockConnection, times(1)).rollback();
```

Comment:
Is this line called? I thought the control flow would end when the `write` above throws? It might be good to do the above assertion with `assertThrows` . You then also can add a testcase to verify that if rollback throws, you get a suppressed exception on the `SQLException` .

Reply:
You make some good points. Here's a counter-proposal:

We put `write()` in a try block, then call `fail()` to remove false positives, catch the `SQLException`, and then check for the suppressed exception. After this try block, we verify that rollback has been called.

Reply:
This is considered older style since it is easy to forget the `fail()` call in the try block. We should have junit 4.13 with `assertThrows` here, and we have used it in other tests - https://github.com/confluentinc/kafka-connect-jdbc/blob/1b47d549951148c5cf8bc2dfcaee85070f342777/src/test/java/io/confluent/connect/jdbc/sink/BufferedRecordsTest.java#L657

Any reason you prefer the older style?

Reply:
No reason, I was looking if you could get the exception from the `assertThrows` and couldn't find it in docs online. This works for me.